### ISSUE TITLE
Should deleteForward change the direction of model#deleteContent() merge blocks?

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-typing#146]
Let's see the two following cases:

TC1: `<h1>Foo</h1><p>[]Bar</p>` + delete (backspace)
TC2: `<h1>Foo[]</h1><p>Bar</p>` + forward delete

Now, we could handle these cases differently. Right now, in both of them the result will be `<h1>Foo[]Bar</h1>`. However, perhaps the user's intention is a bit different when pressing the backward and forward delete keys.

**Comment by scofalik** [comment]
I was actually thinking about the exact same thing like 5 minutes ago. When thinking about scenarios in OT, it just slipped my mind if maybe user has different intentions in those two scenarios. I don't have a case for it in OT, I just thought about it.

Can you give examples of cases where there are two possible intentions? In the given example, I don't see what different effect the user might have wanted to achieve.

**Comment by Reinmar** [comment]
There's a special case reported in https://github.com/ckeditor/ckeditor5-engine/issues/1566. If we'll decide to modify forward delete behaviour by doing "merge right" (as I described in https://github.com/ckeditor/ckeditor5-engine/issues/1566#issuecomment-426200627) we may fix this general issue reported here and the special case reported in https://github.com/ckeditor/ckeditor5-engine/issues/1566.

**Comment by f1ames** [comment]
Checked how this works in Microsoft Word 2016, for both TC1 and TC2 the result is `<h1>Foo[]Bar</h1>`:

![test1 msw](https://user-images.githubusercontent.com/1061942/46340808-cfb3c980-c636-11e8-8ec3-9ca7f885d3b4.gif)

In special case mentioned in https://github.com/ckeditor/ckeditor5-typing/issues/146#issuecomment-426201678, the header remains a header when deleted forward to an empty paragraph:

![test2 msw](https://user-images.githubusercontent.com/1061942/46341662-1276a100-c639-11e8-880d-5be566e5196b.gif)

**Comment by scofalik** [comment]
Yeah when I re- read this issue I actually did not understand what was the problem. Keeping the previous element (header in this case) in all scenarios is the only reasonable expectation.

**Comment by jodator** [comment]
I think that for empty elements it might remove the block instead of merging:

![peek 2019-02-21 16-46](https://user-images.githubusercontent.com/247363/53182255-69394300-35f9-11e9-98ad-4ee8e2d58c54.gif)

The above feels buggy.

**Comment by scofalik** [comment]
Okay, so the behavior here can be as follow:

`<h2>[]</h2><p>Foo bar</p>` -> delete key -> `<h2>Foo bar</h2>`
`<p>[]</p><h2>Foo bar</h2>` -> delete key -> `<h2>Foo bar</h2>`
`<h2></h2><p>[]Foo bar</p>` -> backspace key -> `<p>Foo bar</p>`

Does it make sense? For delete key, it treats paragraph like a default and a heading like something with additional formatting, so heading is preserved. But backspace has a more clear intention that you want to remove the block above.

Of course, this works only when the first element is empty. If the first element is not empty I think that it should keep it's formatting (paragraph / heading) no matter what is the next element and no matter if backspace or delete was used.

**Comment by Reinmar** [comment]
One more TC: https://github.com/ckeditor/ckeditor5/issues/1630

**Comment by Reinmar** [comment]
One more TC: https://github.com/ckeditor/ckeditor5/issues/895
